Assets/UnityVoiceProcessor/VoiceProcessor.cs

```diff
@@ -1,283 +1,338 @@
+        {
+            get
+            {
+                if (_instance == null) FindObjectOfType<VoiceProcessor>();
+                if (_instance == null)
+                {
+                    _instance = new GameObject("Pv.Unity.VoiceProcessor").AddComponent<VoiceProcessor>();
+                    DontDestroyOnLoad(_instance.gameObject);
+                }
+                return _instance;
+            }
+        }
+
+        /// <summary>
+        /// Add a frame listener that will receive audio frames generated by the VoiceProcessor.
+        /// </summary>
+        /// <param name="listener">`VoiceProcessorFrameListener` for processing frames of audio.</param>
+        public void AddFrameListener(VoiceProcessorFrameListener listener)
+        {
+            _onFrame += listener;
+        }
+
+        /// <summary>
+        /// Add multiple frame listeners that will receive audio frames generated by the VoiceProcessor.
+        /// </summary>
+        /// <param name="listeners">`VoiceProcessorFrameListeners` for processing frames of audio.</param>
+        public void AddFrameListeners(VoiceProcessorFrameListener[] listeners)
+        {
+            foreach (var listener in listeners)
+            {
+                _onFrame += listener;
+            }
+        }
+
+        /// <summary>
+        /// Remove a frame listener from the VoiceProcessor. It will no longer receive audio frames.
+        /// </summary>
+        /// <param name="listener">`VoiceProcessorFrameListener` that you would like to remove.</param>
+        public void RemoveFrameListener(VoiceProcessorFrameListener listener)
+        {
+            _onFrame -= listener;
+        }
+
+        /// <summary>
+        /// Remove frame listeners from the VoiceProcessor. They will no longer receive audio frames.
+        /// </summary>
+        /// <param name="listeners">`VoiceProcessorFrameListeners` that you would like to remove.</param>
+        public void RemoveFrameListeners(VoiceProcessorFrameListener[] listeners)
+        {
+            foreach (var listener in listeners)
+            {
+                _onFrame -= listener;
+            }
+        }
+
+        /// <summary>
+        /// Clears all currently registered frame listeners.
+        /// </summary>
+        public void ClearFrameListeners()
+        {
+            _onFrame = null;
+        }
+
+        /// <summary>
+        /// Starts audio recording with the specified audio properties.
+        /// </summary>
+        /// <param name="frameLength">The length of each audio frame, in number of samples.</param>
+        /// <param name="sampleRate">The sample rate to record audio at, in Hz.</param>
+        public void StartRecording(int frameLength, int sampleRate)
+        {
+            if (IsRecording)
+            {
+                // if sample rate or frame size have changed, restart recording
+                if (sampleRate != SampleRate || frameLength != FrameLength)

```

Comment:
so in the VPs that I've currently completed, I've been throwing an error if start is called while recording with different parameters. I think that's better behaviour, because if the user calls start with new params without dealing with previous listeners, those listeners will suddenly receive audio they can't handle. If we throw, we teach the user to cleanup properly between different recording sessions.